working/333 - shared memory multithreading/proposal.md

```diff
@@ -1272,7 +932,7 @@ throw.
 ```dart
```

Comment:
Why can't an `AtomicRef` contains not-real-ref-types? Seems like forcing boxing should be sufficient to make it work. Also, if you can store the value in a singleton unmodifiable list, and store the list in the `AtomicRef`, it feels like a distinctions without a real difference.
At least if we have Java-like memory model semantics for final fields, then:
```dart
class LockBox<T> {
  final T value;
  LockBox(this.value);
}
shared AtomicRef<LockBox<int>> _box = LockBox(0);
int get atomicInt => _box.load().value;
set atomicInt(int value) { _box.store(LockBox(value)); }
```
should give you a guaranteed store of an integer.

Reply:
I wanted to partition types which behave like value types from those that are reference types to make performant implementation simpler. I also want don't boxing and surprising behaviors which can arise from it: consider the semantics of `compareAndSwap` for example which is defined in terms of `identical`.

Reply:
Actually now that I think about it disallowing `AtomicRef<int>` does not disallow `AtomicRef<Object> ref; ref.compareAndSwap(10.0, 20.0)` which requires implementation to peak into the value. I think I need to define that behavior of `compareAndSwap` is undefined for primitives like `int` and `double`.